edu-services/cm-service/cm-impl/hibernate-impl/impl/src/java/org/sakaiproject/coursemanagement/impl/CourseSitePublishServiceImpl.java

```diff
@@ -0,0 +1,244 @@
+    * unpublished.  In order to prevent this case from happening, whenever this service marks a course site as published, it also sets a site property called <i>course_site_publish_service.publish.set</i>.
+    * Then, if an instructor manually resets his site to be unpublished, this service will not automatically publish it.  Rather, the instructor who manually unpublished the
+    * course site will have to go into sakai, use the site setup tool, and manually publish it.
+    * </p>
+    * </br></br>
+    * @param numDaysBeforeTermStarts   number of days before a term starts that course sites should be published.
+    * </br></br>
+    * @return the number of course sites that were published.
+    */
+   public int publishCourseSites(int numDaysBeforeTermStarts) {
+
+      logger.info("publishCourseSites(" + numDaysBeforeTermStarts + " days before the term starts)");
+
+      Date today             = new Date();
+      Date publishDate       = null;
+      int  numSitesPublished = 0;
+
+      try {
+         // get the list of the academic terms
+         List<AcademicSession> academicSessions = courseManagementService.getAcademicSessions();
+
+         for(AcademicSession academicSession : academicSessions) {
+            publishDate = new Date(academicSession.getStartDate().getTime() - numDaysBeforeTermStarts * ONE_DAY_IN_MS);
+            // see if the academic is scheduled to start within the specified number of days
+            if (publishDate.getTime() < today.getTime() && today.getTime() < academicSession.getEndDate().getTime()) {
+
+               // get a list of all published and unpublished course sites in ascending creation date order which are associated with the specified academic session
+               Hashtable<String, String> propertyCriteria = new Hashtable<String, String>();
+               propertyCriteria.put("term_eid", academicSession.getEid());
+               List<Site> sites = (List<Site>)siteService.getSites(SelectionType.NON_USER, "course", null, propertyCriteria, SortType.CREATED_ON_ASC, null);
+
+               for(Site site : sites) {
+                  site.loadAll();
+                  String siteTermEid = site.getProperties().getProperty("term_eid");
+                  if (siteTermEid != null && siteTermEid.length() > 0 && siteTermEid.trim().equals(academicSession.getEid().trim())){
+                  if (!site.isPublished()) {
+                     // see if this service has already published course site once before.
+                     // if it has, then someone has manually reset the published flag, and wants the course to be unpublished.
+                     // so don't switch it back to being published - just leave it as unpublished.
+                     ResourceProperties siteProperties = site.getProperties();
+                     String siteProperty   = siteProperties.getProperty(SITE_PROPERTY_COURSE_SITE_PUBLISHED);
+
+                     if (siteProperty == null) {
```

Comment:
I don't think there is another use for that property, so both options are fine. Maybe yours allow to add other options in the future in that property, so I'll use your solution.